lib/client.dart

```diff
@@ -1,59 +1,128 @@
 import 'dart:async';
 import 'dart:convert';
+
+import 'package:gql/execution.dart' as exec;
+import 'package:gql/language.dart';
+import 'package:gql/link.dart' as gql_link;
 import 'package:http/http.dart' as http;
 import 'package:json_annotation/json_annotation.dart';
+
 import './schema/graphql_error.dart';
 import './schema/graphql_query.dart';
 
-class _DefaultHttpJsonClient extends http.BaseClient {
+/// Simple implementation of an HttpLink
+class SimpleHttpJsonLink implements gql_link.Link {
+  /// Endpoint of a GraphQL service
+  final String graphQLEndpoint;
+
   http.Client _httpClient = http.Client();
 
-  _DefaultHttpJsonClient();
+  /// Construct the Link
+  SimpleHttpJsonLink(this.graphQLEndpoint);
+
+  Future<exec.Response> _post(exec.Request request) async {
+    final dataResponse = await _httpClient.post(
+      graphQLEndpoint,
+      headers: {
+        'Content-type': 'application/json',
+        'Accept': 'application/json',
+      },
+      body: json.encode({
+        'operationName': request.operation.operationName,
+        'variables': request.operation.variables,
+        'query': printNode(request.operation.document),
+      }),
+    );
+
+    final Map<String, dynamic> jsonBody =
+        json.decode(utf8.decode(dataResponse.bodyBytes));
+
+    final response = GraphQLResponse.fromJson(jsonBody);
+
+    // TODO: implement some `GraphQLJsonResponse`
+    // to avoid this re-mapping the response.
+    //
+    // `package:gql:execution` does not care about the
+    // transport protocol and the format. It is `SimpleHttpJsonLink`s
+    // responsibility because it chooses to use Http and Json.
+    return exec.Response(
+      errors: response.errors?.map(
+        (e) => exec.GraphQLError(
+          message: e.message,
+          locations: e.locations?.map(
+            (l) => exec.ErrorLocation(
+              line: l.line,
+              column: l.column,
+            ),
+          ),
+          path: e.queryPath,
+        ),
+      ),
+      data: jsonBody['data'],
+    );
```

Comment:
According to the current spec of GraphQL `errors` may contain an `extensions` key with the additional data on `errors`: https://graphql.github.io/graphql-spec/June2018/#example-fce18 . And we should support.
Also, servers using an older version of the spec may still return extra keys in the same level as the `path`, `message` and `locations` - I think it would be good to not trim those, but we should at least follow the spec.

For me, this is a blocker given that we rely on extension data to handle errors properly

Reply:
That's a great point. I guess, response should also expose the raw response json.

Reply:
@leoiacovini on the second look, `extensions` on `error` are not yet implemented in Artemis. Which just point to the kind of problems that happen when every client implements it's own version of a common feature.

Reply:
Sure thing. I think it would be great to have the `extensions` as a separated property (given that it's part of the spec) but should give access to the raw error map

Reply:
I am hesitant to add the raw "anything". Earlier I wrote "expose the raw response json", but today I realize that's protocol-specific and serialisation-format-specific.

I am now thinking that any non-standard behavior should be implemented by extending the `Response` to contain such non-standard fields. But it also means that the `Link` has to be extended. So that needs to be taken into account when implementing a more feature-ful `HttpLink`.

Reply:
I've now implemented the `HttpLink` in a way that any non-standard fields could be handled by extending base classes.